x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/use_find_rules_query.ts

```diff
@@ -0,0 +1,136 @@
+  queryArgs: FindRulesQueryArgs,
+  queryOptions: UseQueryOptions<
+    RulesQueryData,
+    Error,
+    RulesQueryData,
+    [...string[], FindRulesQueryArgs]
+  >
+) => {
+  const { addError } = useAppToasts();
+
+  return useQuery(
+    [FIND_RULES_QUERY_KEY, ...queryPrefix, queryArgs],
+    async ({ signal }) => {
+      const response = await fetchRules({ signal, ...queryArgs });
+
+      return { rules: response.data, total: response.total };
+    },
+    {
+      refetchIntervalInBackground: false,
+      onError: (error: Error) => addError(error, { title: i18n.RULE_AND_TIMELINE_FETCH_FAILURE }),
+      ...queryOptions,
+    }
+  );
+};
+
+/**
+ * We should use this hook to invalidate the rules cache. For example, rule
+ * mutations that affect rule set size, like creation or deletion, should lead
+ * to cache invalidation.
+ *
+ * @returns A rules cache invalidation callback
+ */
+export const useInvalidateRules = () => {
+  const queryClient = useQueryClient();
+
+  return useCallback(() => {
+    /**
+     * Invalidate all queries that start with FIND_RULES_QUERY_KEY. This
+     * includes the in-memory query cache and paged query cache.
+     */
+    queryClient.invalidateQueries(FIND_RULES_QUERY_KEY, {
+      refetchActive: true,
+      refetchInactive: false,
+    });
+  }, [queryClient]);
+};
+
+/**
+ * We should use this hook to update the rules cache when modifying rules
+ * without changing the rules collection size. Use it with the new rules data
+ * after operations like bulk or single rule edit or rule activation, but not
+ * when adding or removing rules. When adding/removing rules, we should
+ * invalidate the cache instead.
+ *
+ * @returns A rules cache update callback
+ */
+export const useUpdateRulesCache = () => {
+  const queryClient = useQueryClient();
+  /**
+   * Use this method to update rules data cached by react-query.
+   * It is useful when we receive new rules back from a mutation query (bulk edit, etc.);
+   * we can merge those rules with the existing cache to avoid an extra roundtrip to re-fetch updated rules.
+   */
+  return useCallback(
+    (newRules: Rule[]) => {
+      queryClient.setQueriesData<ReturnType<typeof useFindRulesQuery>['data']>(
+        FIND_RULES_QUERY_KEY,
+        (currentData) =>
+          currentData
+            ? {
+                rules: updateRules(currentData.rules, newRules),
+                total: currentData.total,
+              }
+            : undefined
+      );
+    },
+    [queryClient]
+  );
+};
+
+/**
+ * Update cached rules with the new ones
+ *
+ * @param currentRules
+ * @param newRules
+ */
+export function updateRules(currentRules: Rule[], newRules: Rule[]): Rule[] {
+  const currentRuleIds = new Set(currentRules.map((r) => r.id));
+
+  return newRules.reduce(
+    (updatedRules, newRule) =>
+      currentRuleIds.has(newRule.id)
+        ? updatedRules.map((rule) => (newRule.id === rule.id ? newRule : rule))
+        : updatedRules,
+    currentRules
```

Comment:
We can improve it by using map of new rules.
This way you won't map through the whole updateRules/currentRules array every reduce callback, we'll just map currentRules once

```
  const newRulesMap = new Map(newRules.map((rule) => [rule.id, rule));

  return currentRules.map((rule) => newRulesMap.get(rule.id) ?? rule);
```

Reply:
If the cached and newly fetched rules have no intersections, I want to keep referential equality. But in your snippet, all cache entries get updated unconditionally.

Reply:
we can do something like this, if this is an objective to keep referential equality. 

```typescript

  const newRulesMap = new Map(newRules.map((rule) => [rule.id, rule]));

  if (!currentRules.some((rule) => newRulesMap.has(rule.id))) {
    return currentRules;
  }
   
  return currentRules.map((rule) => newRulesMap.get(rule.id) ?? rule);
```

Reply:
Yes, it could be done as you propose. I don't have a strong opinion on either option, to be honest. Do you consider it a blocker for this PR?

Reply:
I think if we have an opportunity to do something in more efficient way, why don't we then?
Any concerns regarding proposed code?

I did a quick benchmarking

3k rules mock:
```typescript
const oldRules = Array.from({ length: 10 * 300 }).map((id) => ({ id }));
const newRules = Array.from({ length: 10 }).map((id) => ({ id }));

const updateLinear = (currentRules, newRules) => {
  const newRulesMap = new Map(newRules.map((rule) => [rule.id, rule]));

  if (!currentRules.some((rule) => newRulesMap.has(rule.id))) {
    return currentRules;
  }

  return currentRules.map((rule) => newRulesMap.get(rule.id) ?? rule);
};

const updateQuadratic = (currentRules, newRules) => {
  const currentRuleIds = new Set(currentRules.map((r) => r.id));

  return newRules.reduce(
    (updatedRules, newRule) =>
      currentRuleIds.has(newRule.id)
        ? updatedRules.map((rule) => (newRule.id === rule.id ? newRule : rule))
        : updatedRules,
    currentRules
  );
};
const tq = performance.now();
updateQuadratic(oldRules, newRules);
console.log('tq', performance.now() - tq);

const tl = performance.now();
updateLinear(oldRules, newRules);
console.log('tl', performance.now() - tl);
```
```bash
tq 1.5999999940395355
tl 0.30000001192092896
```

Doesn't seem a lot, but it's measurable and it's my fast laptop(if you consider re-render time 'limit' ~ 16ms, that would even make more sense)
I'm not keen of optimizations for just optimizations, but this looks to me like an easy pick, with no need to invest any time for us.

Reply:
Well, in the case of 20 rules per page and only one rule updated (probably 99% of use cases), the "quadratic" (actually n*m) implementation turns out to be a little bit faster:

```
tq 0.09999999403953552
tl 0.10000002384185791
```

Micro optimizations could have a nonobvious effect, which is sometimes the opposite of what you'd expect. And it is pretty hard to measure their actual impact on performance. That's why I prefer to avoid them unless there is a strong reason. And the `1ms` difference in performance in very rare cases doesn't seem like a good reason.

That said, if for you it's a blocker, let's merge your snippet and move on.

Reply:
I think your result is one time fluctuation, that could happen only because of a little data set and micro time. In some runs I can see quadratic doing better, in some worse

if I loop it 100 times, result would be like this:
```
const tq = performance.now();
for(let i =0; i < 100; i ++)
updateQuadratic(oldRules, newRules);
console.log('tq', performance.now() - tq);
const tl = performance.now();
for(let i =0; i < 100; i ++)
updateLinear(oldRules, newRules);
console.log('tl', performance.now() - tl);

```

```
tq 0.4000000059604645
tl 0.30000001192092896
```
with a linear way performing better for a case 1 of 20, in most of the runs.


> Micro optimizations could have a nonobvious effect, which is sometimes the opposite of what you'd expect. And it is pretty hard to measure their actual impact on performance. That's why I prefer to avoid them unless there is a strong reason. 

I agree with you on that and it's not my goal to chase microseconds win in performance. It's just it looks as an easy win to map through potentially large array in linear time not a quadratic.